fenicsadapter/fenicsadapter.py

```diff
@@ -1,759 +1,458 @@
-        """Sets the coupling mesh. Called by initalize() function at the
-        beginning of the simulation.
-        """
-        self._coupling_subdomain = subdomain
-        self._mesh_fenics = mesh
-        self._fenics_vertices, self._coupling_mesh_vertices, self._n_vertices = self._extract_coupling_boundary_vertices()
-        self._vertex_ids = self._interface.set_mesh_vertices(self._mesh_id, self._coupling_mesh_vertices)
-
-        """ Define a mapping between coupling vertices and their IDs in precice"""
+        """ Set up mesh in preCICE """
+        self._vertex_ids = self._interface.set_mesh_vertices(self._interface.get_mesh_id(
+            self._config.get_coupling_mesh_name()), self._coupling_mesh_vertices)
+
+        """ Define a mapping between coupling vertices and their IDs in preCICE """
         id_mapping = dict()
         for i in range(self._n_vertices):
-            id_mapping[self._fenics_vertices[i].global_index()] = self._vertex_ids[i]
-
-        if use_nearest_projection:
-            self._edge_vertex_ids1, self._edge_vertex_ids2 = self._extract_coupling_boundary_edges(id_mapping)
-            for i in range(len(self._edge_vertex_ids1)):
-                assert(self._edge_vertex_ids1[i] != self._edge_vertex_ids2[i])
-                self._interface.set_mesh_edge(self._mesh_id, self._edge_vertex_ids1[i], self._edge_vertex_ids2[i])
+            id_mapping[fenics_vertices[i].global_index()] = self._vertex_ids[i]
+        edge_vertex_ids1, edge_vertex_ids2 = get_coupling_boundary_edges(mesh, coupling_subdomain, id_mapping)
 
-    def _set_write_field(self, write_function_init):
-        """Sets the write field. Called by initalize() function at the
-        beginning of the simulation.
+        """ Set mesh edges in preCICE to allow nearest-projection mapping"""
+        for i in range(len(edge_vertex_ids1)):
+            assert (edge_vertex_ids1[i] != edge_vertex_ids2[i])
+            self._interface.set_mesh_edge(self._interface.get_mesh_id(self._config.get_coupling_mesh_name()),
+                                          edge_vertex_ids1[i], edge_vertex_ids2[i])
 
-        :param write_function_init: function on the write field
-        """
-        self._write_function_type = determine_function_type(write_function_init)
-        self._write_data = self._convert_fenics_to_precice(write_function_init)
+        """ Set read functionality parameters """
+        self._read_function_type = determine_function_type(read_function)
+        self._function_space = function_space
 
-    def _set_read_field(self, read_function_init):
-        """Sets the read field. Called by initalize() function at the
-        beginning of the simulation.
+        return self._interface.initialize()
 
-        :param read_function_init: function on the read field
+    def initialize_data(self, write_function):
```

Comment:
`write_function=None` should also be a valid input. Example use cases: One-way-coupling or if only one participant has `initialize=True`.